### ISSUE TITLE
ClassDiscriminatorMode annotation for data classes

### ISSUE THREAD

**Issue by rattletat** [Kotlin/kotlinx.serialization#2883]
Hi there,

similarly to [JsonClassDiscriminator](https://kotlinlang.org/api/kotlinx.serialization/kotlinx-serialization-json/kotlinx.serialization.json/-json-class-discriminator/), I would like an annotation type to indicate the discriminator mode for data classes:

```kotlin
@Serializable
@JsonClassDiscriminatorMode(ClassDiscriminatorMode.ALL_JSON_OBJECTS)
abstract class Base
```

The motivation is that we have a polymorphic data class consumed by third-party actors but we want to ensure that the serialized classes always include the class type annotation, regardless of whether the third-party actors serialize it using the concrete or the polymorphic type.

If there is a work-around already known, please feel free to share.

**Comment by sandwwraith** [comment]
Why you need annotation exactly and why the general `Json` setting does not satisfy your needs?

**Comment by rattletat** [comment]
The use-case here is that we define message types consumed by multiple parties who should be able to create and serialize messages. These created messages (JSONs) will be processed by us again, for which we need to deserialize them and rely on the type field. We provide functions to serialize and deserialize them correctly (using the polymorphic type), but it is easy for the third-parties to mess up here:

```
# This will work
Json.encodeToString(someConcreteMessage as Message)
```
```
# This will fail
Json.encodeToString(subscriptionMessage)
```

Having an annotation directly on the data class to indicate to always include the type field would decrease the risk of some party not including the type field massively.

Is there a workaround to achieve something similar? I would suspect that one could include the serialName directly into the data class, but not sure if that would play nicely with the library.

**Comment by rattletat** [comment]
The use-case here is that we define message types consumed by multiple parties who should be able to create and serialize messages. These created messages (JSONs) will be processed by us again, for which we need to deserialize them and rely on the type field. We provide functions to serialize and deserialize them correctly (using the polymorphic type), but it is easy for the third-parties to mess up here:

```
# This will work
Json.encodeToString(someConcreteMessage as Message)
```
```
# This will fail
Json.encodeToString(someConcreteMessage)
```

Having an annotation directly on the data class to indicate to always include the type field would decrease the risk of some party not including the type field massively.

Is there a workaround to achieve something similar? I would suspect that one could include the serialName directly into the data class, but not sure if that would play nicely with the library.

**Comment by sandwwraith** [comment]
Hm, I see. So `Json` will be created by 3rd party, and therefore, you don't control `JsonBuilder.classDiscriminatorMode`? 
The problem with specifying the wrong type is well-known to us, and we are planning to eventually add IDE inspection for that.

**Comment by rattletat** [comment]
Exactly. We provide methods on the Base class to do the serialization correctly i.e. `someConcreteMessage.toJson()`, but our clients could just ignore this serialization method and call `Json.encodeToString(someConcreteMessage)` because they saw it in some tutorial^^ In that case, setting `ClassDiscriminatorMode.ALL_JSON_OBJECTS` on the data class would still include the type discriminator in the resulting JSON.
Do you have an idea how this could be accomplished in other ways? Is there a way to specify a default `Json` serialization object on dataclasses directly?
